sdks/java/io/hadoop-input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/unit/tests/HadoopInputFormatIOTest.java

```diff
@@ -0,0 +1,796 @@
+    PAssert.that(actual).containsInAnyOrder(expected);
+    p.run();
+  }
+
+  /**
+   * This test validates behavior of {@link HadoopInputFormatBoundedSource} if RecordReader object creation fails
+   * in {@link HadoopInputFormatBoundedSource.HadoopInputFormatReader#start() start()} method.
+   */
+  @Test
+  public void testReadersStartIfCreateRecordReaderFails() throws Exception {
+    List<BoundedSource<KV<Text, Employee>>> boundedSourceList =
+        getBoundedSourceList(BadCreateReaderInputFormat.class, Text.class, Employee.class,
+            WritableCoder.of(Text.class), AvroCoder.of(Employee.class));
+    for (BoundedSource<KV<Text, Employee>> source : boundedSourceList) {
+      BoundedReader<KV<Text, Employee>> reader = source.createReader(p.getOptions());
+      thrown.expect(Exception.class);
+      thrown
+          .expectMessage("Exception in creating RecordReader in BadCreateRecordReaderInputFormat");
+      reader.start();
+    }
+  }
+
+  /**
+   * This test validates behavior of HadoopInputFormatSource if
+   * {@link InputFormat#createRecordReader() createRecordReader()} of InputFormat returns null.
+   */
+  @Test
+  public void testReadersStartWithNullCreateRecordReader() throws Exception {
+    List<BoundedSource<KV<Text, Employee>>> boundedSourceList = getBoundedSourceList(
+        BadNullCreateReaderInputFormat.class,
+        Text.class,
+        Employee.class,
+        WritableCoder.of(Text.class),
+        AvroCoder.of(Employee.class));
+    for (BoundedSource<KV<Text, Employee>> source : boundedSourceList) {
+      BoundedReader<KV<Text, Employee>> reader = source.createReader(p.getOptions());
+      thrown.expect(IOException.class);
+      thrown.expectMessage(
+          String.format(HadoopInputFormatIOContants.NULL_CREATE_RECORDREADER_ERROR_MSG,
+              new BadNullCreateReaderInputFormat().getClass()));
+      reader.start();
+    }
+  }
+
+  /**
+   * This test validates behavior of {@link HadoopInputFormatBoundedSource#createReader()
+   * createReader()} and {@link HadoopInputFormatBoundedSource.HadoopInputFormatReader#start()
+   * start()} methods if InputFormat's {@link InputFormat#getSplits() getSplits()} returns
+   * InputSplitList having having no records.
+   */
+  @Test
+  public void testReadersCreateReaderAndStartWithZeroRecords() throws Exception {
+    List<BoundedSource<KV<Text, Employee>>> boundedSourceList = getBoundedSourceList(
+        BadNoRecordsInputFormat.class,
+        Text.class,
+        Employee.class,
+        WritableCoder.of(Text.class),
+        AvroCoder.of(Employee.class));
+    for (BoundedSource<KV<Text, Employee>> source : boundedSourceList) {
+      BoundedReader<KV<Text, Employee>> reader = source.createReader(p.getOptions());
+      assertEquals(false, reader.start());
+    }
+  }
+
+  /**
+   * This test validates fractions consumed while reading data.
+   */
+  @Test
+  public void testReadersGetFractionConsumed() throws Exception {
```

Comment:
We have a test to check if start call returns false, added one to check if getFractionConsumed() returns 1 -testReadersStartWithZeroRecords()